### ISSUE TITLE
Remove duplicate suggestions

### ISSUE THREAD

**Issue by tDeContes** [rugk/unicodify#93]
## Background
It would be nice that the context menu doesn't suggest useless items. (I'm using only casing.)
For ex when I ask for `i`, it suggests `i`/`I`/`I`/`I`.

## Proposed solution
For each suggested text transformation, after the transformation is made, check whether:
- it's the same as original,
- it's already suggested upper.

If yes, remove the item.

It would also be useful for code casing, in all cases where is there only one word to handle.
I don't think it would be useful for those with special characters.

## Alternatives
It could be seen as an alternative to #30, although it is more of a complement.

**Comment by tDeContes** [comment]
> * it's the same as original,
> 
> * it's already suggested upper.
> 
> If yes, remove the item.

If it's the same as original and it's not already suggested upper, it could be inactivated (not selectable) instead of removed.

---

Currently:
![image](https://github.com/user-attachments/assets/3dc10a89-63ec-4bb5-8a78-d0ed6ff8ff79)

It could be reduced to:
![image](https://github.com/user-attachments/assets/08c3fa66-0486-4104-b6b5-66b90accd917)
(with `item` grayed)

**Comment by tdulcet** [comment]
> If it's the same as original and it's not already suggested upper, it could be inactivated (not selectable) instead of removed.

Interesting, this is a lot easier to implement than removing the menu items, although there would still be duplicates:
![image](https://github.com/user-attachments/assets/fe16510b-b6c1-4626-897c-05b7dd2fbbf2)
I included this change in #96. Let me know if it would satisfy your request.

**Comment by tDeContes** [comment]
> > If it's the same as original and it's not already suggested upper, it could be inactivated (not selectable) instead of removed.
> 
> Interesting, this is a lot easier to implement than removing the menu items, although there would still be duplicates

Why is it difficult to remove the menu items?
Annoying: duplicates are the most important.

> I included this change in #96. Let me know if it would satisfy your request.

Anyway it's an improvement. Keep it! :-)

**Comment by tdulcet** [comment]
> Why is it difficult to remove the menu items?

Removing the menu items is not difficult per say, it is determining which items to remove. By default, Unicodify provides a nested context menu with descriptions of each menu item:
![image](https://github.com/user-attachments/assets/7e15c631-2d40-444a-9584-2da6083bff1d)
It would likely be confusing for users if they had to go into a different submenu and select a different menu item then what they actually wanted in order to get the desired conversion/casing.

> Anyway it's an improvement. Please keep it! :-)

No problem, I will keep this change in #96 and it should be included in the next update.

**Comment by tDeContes** [comment]
> > Why is it difficult to remove the menu items?
> 
> Removing the menu items is not difficult per say, it is determining which items to remove.

If I said "So, disable them instead.", would it be the same difficulty, or is it a UI pb only?

> By default, Unicodify provides a nested context menu with descriptions of each menu item

> It would likely be confusing for users if they had to go into a different submenu and select a different menu item then what they actually wanted in order to get the desired conversion/casing.

I might suggest making a setting to allow users to choose that, but maybe the `Show a readable description of the transformation` setting is enough: if checked, users need all items to avoid any confusion; if not, it's useless (IMHO even with the nested context menu).

**Comment by tdulcet** [comment]
> If I said "So, disable them instead.", would it be the same difficulty, or is it a UI pb only?

One would not want to remove the menu items per say, as there is a performance cost to add them back. Instead, one can ether disable or hide each of them, both of which are equally easy to do. The problem is then determining which ones to hide. The way the add-on is currently designed, each menu item is handled independently, with no knowledge of the other items.

**Comment by tDeContes** [comment]
> > If I said "So, disable them instead.", would it be the same difficulty, or is it a UI pb only?

> The problem is then determining which ones to hide. The way the add-on is currently designed, each menu item is handled independently, with no knowledge of the other items.

OK, so it would be the same difficulty, because it's not sequential.

Well, what would be some ways to make each menu item know the value of those above it, but not those below it?
Maybe a shared array, with each menu item knowing its position on it? (hope not too high cost!)

**Comment by tdulcet** [comment]
> Well, what would be some ways to make each menu item know the value of those above it, but not those below it?

Yeah, we would either need to iterate through the conversions beforehand to determine which ones to make visible or maintain some shared state while updating them. This is something we can consider for a future update.

The question is whether the duplicates should be removed/hidden per category (the four listed on the options page) or for all of the conversions combined. In the latter case, those separator lines should probably be removed as well.

There is also the question of which combination of options should enable this feature. I would say the "readable description" should be off, while the "live preview" should be on. The "nested context menu" option should probably be off as well, unless maybe if we were hiding the duplicates per category.

**Comment by tDeContes** [comment]
> > Well, what would be some ways to make each menu item know the value of those above it, but not those below it?
> 
> Yeah, we would either need to iterate through the conversions beforehand to determine which ones to make visible or maintain some shared state while updating them. This is something we can consider for a future update.

I hope it will be soon. :-)
Unfortunately I don't know JavaScript, so I can't make PR.

But maybe I can help you with the algorithm?
What are the advantages and disadvantages of each?

When my computer is slow I see that menu items can be updated after the menu is open. Is it the same for hiding?
Can you be aware when the last menu item has finished?
If yes, when they have finished, you can launch an iterative process which will tell each menu item if it has to be hidden.

> 
> The question is whether the duplicates should be removed/hidden per category (the four listed on the options page) 

AFAIK, duplicates cannot happen for the 2 first categories, am I wrong?
If I'm right, you can care about duplicates only for the 2 last categories.

> or for all of the conversions combined. 

I don't need to have each of them twice, and what is obvious is that no one needs duplicates in the same sub-menu when they have the same title.

So if you want to make the best extension possible, and with the mindset that if `readable description` is checked, users need all items to avoid any confusion, it will be the same about nested context menu.

> In the latter case, those separator lines should probably be removed as well.

Still if you want to make the best extension possible, yes, it would be better. :-)
(Keep me aware of the constraints, so I can see how I can help you.)

> 
> There is also the question of which combination of options should enable this feature. I would say the "readable description" should be off, while the "live preview" should be on. 

You're true: We need to check both, since when they're both unchecked `readable description` preempts.

> The "nested context menu" option should probably be off as well, unless maybe if we were hiding the duplicates per category.


---

> Disable duplicates handling
> 
> If `live preview` and not `readable description` then:
> 
> > Enable duplicates handling
> > 
> > Case `nested context menu`:
> > 
> > Checked:
> > 
> > > Reset the duplicates handling process for each sub-menu
> > 
> > Unchecked:
> > 
> > > Reset the duplicates handling process only once at the beginning

---

Have you recently checked the case where `readable description` is checked and `live preview` is unchecked?
The readable description appears twice: once normal and once transformed. Is this a bug?

When they're both unchecked, the readable description appears transformed. Maybe this is a bug too, but I like it. :-)

**Comment by tdulcet** [comment]
> When my computer is slow I see that menu items can be updated after the menu is open. Is it the same for hiding?

Every time the context menu is opened on an editable context and the "live preview" option is enabled, all of the menu items are updated. It does not matter how they are updated (hidden, disabled, changing the title, etc.), it is all the same speed.

> Can you be aware when the last menu item has finished?

Yes, but the updating is slow (requires calling an async Web Extension API), so one would not want to wait for that to complete or do it more than once per menu item.

> AFAIK, duplicates cannot happen for the 2 first categories, am I wrong?

The Unicode fonts only support ASCII characters and usually only a small subset, so for example if the user selects text without any  Latin letters, there would be conversions that would not make any changes to the text and thus there would be duplicate menu items. Here are the supported characters for each Unicode font (note that ASCII characters are used as placeholders missing ones): https://github.com/rugk/unicodify/blob/374ed9140e6545b419d8b56a07dec25fa2c9852d/src/common/modules/data/Fonts.js#L166-L192

> Still if you want to make the best extension possible, yes, it would be better. :-)

Yes, we always want to make the best extension possible! 🙂 Of course, that has to be balanced with how many users might use this functionality and the additional complexity. Because the readable description option is enabled by default, I am not sure how many users would explicitly disable it.

> The readable description appears twice: once normal and once transformed. Is this a bug?

No, that is the expected behavior.

> When they're both unchecked, the readable description appears transformed. Maybe this is a bug too, but I like it. :-)

Yeah, this was the original behavior I implemented. The "readable description" and "live preview" options were added later by @rugk.

Thanks for that flow chart of your desired behavior with the various combinations of options. That should be very helpful for if/when someone implements this feature.

**Comment by rugk** [comment]
I don't see a flow chart, where is a flow chart?

So from what I read, whether we disable/hide the menu item does not matter much, but actually this can result in UX issues and needs to consider all different options of the extension in mind. 🙃 

Much has been said, but we would indeed need to think through all options, a flow-chart indeed sounds very useful.
And I agree, we must avoid:
* (1) users not finding "their" option, because it is hidden somewhere
* (2) users being confused, because an option appears multiple times (what this issue is about)

So this is kind of an UX tradeoff as both aims seem to contradict each other.

in the essence, AFAIK, when you see a "readable description" we should allow any duplicates. Only when you only(!) see the live preview, this algorithm could apply.
And in such a case, we could actually hide the items. _Except_ if they are in sub-categories, because then we would violate the aim (1). Yet again, except if that is inside the same sub-menu. That would be fine.
And finally, yes, I agree separators also need to be handled. 

The idea sounds easy, the implementation does sounds way less easier if you think about all this…

## Implementation

As for an implementation hint, here is the relevant code (actually the whole file more or less, but this is what matters):
https://github.com/rugk/unicodify/blob/374ed9140e6545b419d8b56a07dec25fa2c9852d/src/background/modules/ContextMenu.js#L123

As you can see, it _already_ is quite complex, so making it even more complex sounds fun… 😉

**Comment by tdulcet** [comment]
> I don't see a flow chart, where is a flow chart?

@tDeContes used the Markdown quote syntax, so it is somewhat hidden at first glance: https://github.com/rugk/unicodify/issues/93#issuecomment-2593720655. Converting it to a list may be more readable, although GitHub does support [creating actual flow charts](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams):
1. Disable duplicates handling 
2. If `live preview` and not `readable description` then: 
    1. Enable duplicates handling
    2. Case `nested context menu`:
    3. Checked:
        1. Reset the duplicates handling process for each sub-menu
    4. Unchecked:
        1. Reset the duplicates handling process only once at the beginning

**Comment by tDeContes** [comment]
> > I don't see a flow chart, where is a flow chart?
> 
> [@tDeContes](https://github.com/tDeContes) used the Markdown quote syntax, 

Sorry, what I was looking for was a code block, and the quote block was the closest thing I could find in the toolbar buttons.

```
Disable duplicates handling
If `live preview` and not `readable description` then:
  Enable duplicates handling
  Case `nested context menu`:
  Checked:
    Reset the duplicates handling process for each sub-menu
  Unchecked:
    Reset the duplicates handling process only once at the beginning
```


> Converting it to a list may be more readable, 

You're true, lists are a good tool for that. :-)

1. Disable duplicates handling
2. If `live preview` and not `readable description` then:
    - Enable duplicates handling
    - Case `nested context menu`:
        - Checked:
            Reset the duplicates handling process for each sub-menu
        - Unchecked:
            Reset the duplicates handling process only once at the beginning


> although GitHub does support [creating actual flow charts](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams):

Mermaid seems to be broken now (`Unable to render rich display`) but if it's repaired later and you find it funny, I could try it. :-)

**Comment by Thomas-D-C** [comment]
```mermaid
graph TD;

    rd([Readable description])
    lp([Live preview])
    ncm([Nested context menu]);

    rd-c{Checked ?}
    lp-c{Checked ?}
    ncm-c{Checked ?}
    not1(NOT)
    and1(AND)
%%    and2{AND}

    Disable[Disable duplicates handling];
    Enable[Enable duplicates handling];
    rem[Reset the duplicates handling process for each sub-menu];
    rob[Reset the duplicates handling process only once at the beginning];


rd --> rd-c
lp --> lp-c
ncm --> ncm-c
rd-c -- Yes --> Disable
rd-c -- No --> not1
lp-c -- No --> Disable
lp-c -- Yes --> and1

not1 -- Yes --> and1
and1 -- No --> Disable
and1 -- Yes --> Enable

Enable -- Yes --> ncm-c

ncm-c -- Yes --> rem
ncm-c -- No --> rob

```

Well, I tried Mermaid to create a flow chart.
It looks powerful, but I don't know how to use it! :-D
(For example, I think I'm mixing up Grafcet diagrams and logical operators.)

If you know, please feel free to help me. :-)

**Comment by Thomas-D-C** [comment]
> The way the add-on is currently designed, each menu item is handled independently, with no knowledge of the other items.

If I'm not mistaken, it's not sequential but not parallel either, it's just encapsulation?


> we would either need to iterate through the conversions beforehand to determine which ones to make visible or maintain some shared state while updating them.

> > Can you be aware when the last menu item has finished?
> 
> Yes, but the updating is slow (requires calling an async Web Extension API), so one would not want to wait for that to complete or do it more than once per menu item.

Well, IMHO the best would be to use [Sets](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) (one global + one for each sub-menu).

**Comment by Thomas-D-C** [comment]
> It looks powerful, but I don't know how to use it! :-D
> (For example, I think I'm mixing up Grafcet diagrams and logical operators.)

I think this one is a bit better (closer to Grafcet diagrams).

```mermaid
graph TD;

    Start((Start))

    lp{Live preview}
    rd{Readable description}

    Disable([Disable duplicates handling])
    Enable[[Enable duplicates handling]]

subgraph dh[Duplicates handling]
direction TB

    Start-dh((Start))

    ncm{Nested context menu}

    rem([Reset the duplicates handling process for each sub-menu])
    rob([Reset the duplicates handling process only once at the beginning])
end


Start --> lp

lp -- Checked --> rd
lp -- Unchecked --> Disable
rd -- Checked --> Disable
rd -- Unchecked --> Enable

Enable -...-> dh

Start-dh --> ncm

ncm -- Checked --> rem
ncm -- Unchecked --> rob

```

**Comment by Thomas-D-C** [comment]
> > If it's the same as original and it's not already suggested upper, it could be inactivated (not selectable) instead of removed.
> 
> Interesting, this is a lot easier to implement than removing the menu items, although there would still be duplicates 

> I included this change in [#96](https://github.com/rugk/unicodify/pull/96). Let me know if it would satisfy your request.

It seems you are using an API that keeps only the first 100 characters before comparing the text from before and after the transformation.
And it seems that it's not the [`!==` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Strict_inequality) [here](https://github.com/rugk/unicodify/blob/52016320aace932e10a2ee2faf08d792c31e98e7/src/background/modules/ContextMenu.js#L190).

(Do you need a new issue for that?)

**Comment by tdulcet** [comment]
> (Do you need a new issue for that?)

This is a performance optimization, since the context menu generally cannot show more than 100 characters, so when using the live preview option, the text is currently truncated before it is used: https://github.com/rugk/unicodify/blob/5901eb66a1c17dfb5a35780a892b066423dd7921/src/background/modules/ContextMenu.js#L14 Feel free to open a new issue if you believe this should be changed. Do you have a use case that would require comparing more than 100 characters?
